Fix Linux support

CodingKoopa opened the pull request:
This PR fixes some things so that this program can be compiled on Linux using `msbuild` from Mono, or `dotnet build` from the .NET SDK, and then ran using Mono.

![image](https://user-images.githubusercontent.com/13039555/109410692-b0860700-796a-11eb-8871-dc3ac879e04f.png)

The fixes are:
- Removing `é` because of it causing a compile error.
- Making resource paths case sensitive.
- Adding references to `System.Resources.Extensions` because of this error:
```
/usr/share/dotnet/sdk/6.0.100-preview.1.21103.13/Microsoft.Common.CurrentVersion.targets(3136,5): error MSB3822: Non-string resources require the System.Resources.Extensions assembly at runtime, but it was not found in this project's references. [pkNX/pkNX.Sprites/pkNX.Sprites.csproj]
```
In order to do this, it seems to have been necessary to bump the .NET target version to 4.6.1.
- Adding `GenerateResourceUsePreserializedResources` because of this error:
```
/usr/share/dotnet/sdk/6.0.100-preview.1.21103.13/Microsoft.Common.CurrentVersion.targets(3136,5): error MSB3823: Non-string resources require the property GenerateResourceUsePreserializedResources to be set to true. [pkNX/pkNX.Sprites/pkNX.Sprites.csproj]
```
- Preferring the system `flatc` so that the native Linux binary will be used if installed.

Known issues/concerns:
- My knowledge of C# build systems is rather limited, so I'm not sure how correct some of these changes are.
- The _Wild Editor_ is very broken. Sprites do not load, and switching between different options at the top causes a `StackOverflowException`.
- This PR has unnecessary diffs at the top of some files that I haven't been able to get rid of.

ColonelThirtyTwo commented:
How did you get this to build on Linux? Having issues with dotnet build saying that the windows forms libraries are not available.

CodingKoopa commented:
Ah, indeed I never documented what I did. I've went through both of the two methods of compiling pkNX, and fought with them until they both work. For you, I would recommend installing the .NET SDK preview if you can, that's what fixed that issue for me when it comes to building with `dotnet build`.

## General Tips
- If possible, don't have both Mono and the .NET Core SDK installed at the same time - they can interact in weird and unwanted ways.
- Deleting the `pkNX` directory and cloning it once again may be helpful at times.

## Building with the .NET Core SDK
1. Install the [.NET Core SDK](https://dotnet.microsoft.com/download). You may install this using your distribution's package manager, or with the [`dotnet-install.sh`](https://dot.net/v1/dotnet-install.sh) script provided by the .NET Foundation.
**As of writing, the latest stable .NET Core SDK (5.x) doesn't seem to properly provide the `Microsoft.NET.Sdk.WindowsDesktop` SDK. It was necessary for me to install the .NET Core 6.0 SDK preview to successfully compile.**
2. Direct msbuild to the SDK locations (see [here](https://github.com/OmniSharp/omnisharp-roslyn/issues/1394) and [here](https://wiki.archlinux.org/title/.NET_Core#SDK_specified_could_not_be_found) for the origins of this workaround). If you only have one .NET Core SDK installed and don't have Mono installed, this isn't necessary, but it doesn't seem to hurt in any case:
```sh
export MSBuildSDKsPath=/usr/share/dotnet/sdk/$(dotnet --version)/Sdks
```
3. Build the program:
```sh
dotnet build
```

### Troubleshooting

#### .NET Desktop SDK not found
```
error MSB4236: The SDK 'Microsoft.NET.Sdk.WindowsDesktop' specified could not be found.
```
Strangely, this issue I fixed by switching to the .NET SDK 6.0 preview.

## Building with Mono
1. Install the [Mono](https://www.mono-project.com/) framework from your distribution's package manager.
**As of writing, the version of the [Roslyn](https://github.com/dotnet/roslyn) compile included in latest stable Mono release (6.12.0) does not support [C# 9](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-9) - it was necessary for me to install the latest Git version of Mono, although the nightly, alpha, or beta versions may work too.**
2. Install the [Mono fork of msbuild](https://github.com/mono/msbuild) from your distributions package manager. The [Arch Linux package](https://archlinux.org/packages/community/x86_64/mono-msbuild/) refers to this as the "Xamarin implementation" of msbuild, and retrieves it from [here](https://github.com/mono/linux-packaging-msbuild).
3. Restore the Nuget dependencies and build the program, from the root of the repository. This certainly won't work right away - see the troubleshooting section for steps for making the build work.
```sh
msbuild /r
```

### Troubleshooting

#### Too old Mono Version
```
CSC : error CS1617: Invalid option '9' for /langversion. Use '/langversion:?' to list supported values.
```
Your Mono version isn't recent enough.

#### Too many open files
```
Cannot write to the output file "/home/koopa/Fountain/Documents/Projects/C#/pkNX/pkNX.Sprites/obj/Debug/net461/pkNX.Sprites.Properties.Resources.resources". Too many open files
```
I worked around this by raising the open file limit by running `ulimit -n 70000`.

#### .NET Desktop SDK not found
```
error MSB4236: The SDK 'Microsoft.NET.Sdk.WindowsDesktop' specified could not be found.
```
```
warning NETSDK1107: Microsoft.NET.Sdk.WindowsDesktop is required to build Windows desktop applications. 'UseWpf' and 'UseWindowsForms' are not supported by the current SDK.
```
`Microsoft.NET.Sdk.WindowsDesktop` is a type of [.NET project SDK](https://docs.microsoft.com/en-us/dotnet/core/project-sdk/overview), specifically the [.NET Desktop SDK](https://docs.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props-desktop), which includes WinForms. Mono has an implementation of [WinForms](https://en.wikipedia.org/wiki/Windows_Forms), and for that reason I feel that it *should* be able to provide this. I couldn't get this error to go away, though. The maintainer of the [pkHex AUR package](https://aur.archlinux.org/packages/pkhex/) has this same issue, and they never seemed to be able to work around it either. All in all, this feels like a bug or shortcoming in Mono to me. I did the bare minimum to get the project to compile without referencing this SDK.

Strangely, even when the error is resolved, but the warning is still outputted, the build process still can output a working Windows desktop application.

`pkNX.WinForms/pkNX.WinForms.csproj`
```diff
- <Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
+ <Project Sdk="Microsoft.NET.Sdk">
```
```diff
   <ItemGroup Condition="'$(TargetFramework)' == 'net46'">
     <Reference Include="System.Configuration" />
   </ItemGroup>

+  <ItemGroup Condition="'$(TargetFramework)' == 'net461'">
+    <Reference Include="System.Windows.Forms" />
+  </ItemGroup>
```
#### Reference assemblies for .NET 5.0 not found
```
error MSB3971: The reference assemblies for ".NETFramework,Version=v5.0" were not found. You might be using an older .NET SDK to target .NET 5.0 or higher. Update Visual Studio and/or your .NET SDK.
```
I don't really know what the _right_ solution to this would be, but I fixed this by hardcoding the .NET Framework version to one that seems to work.
`pkNX.Tests/pkNX.Tests.csproj`
```diff
-    <TargetFramework>net5.0</TargetFramework>
+    <TargetFramework>net461</TargetFramework>
```
`pkNX.Sprites/pkNX.Sprites.csproj`
```diff
-    <TargetFrameworks>net5.0;net461</TargetFrameworks>
+    <TargetFramework>net461</TargetFramework>
```

## Usage
The built EXE, `pkNX.WinForms/bin/Debug/net461/pkNX.exe`, can be run with:
- Mono for Linux (this is the only one I've tested).
- [Mono for Windows, running in Wine](https://wiki.winehq.org/Mono).
- [Microsoft .NET Framework 4.7.2, running in Wine](https://wiki.winehq.org/Mono#Microsoft_.NET).

ColonelThirtyTwo commented:
@CodingKoopa Thank you for the writeup.
